MAINT: Ensure array_dealloc does not modify refcount of self

seberg opened the pull request:
This is even uglier then before in some ways, but avoids changing
the global totalrefcount in python debug builds.

---

Mostly for discussion. We probably do not want this, OTOH, not using INCREF is necessary if we want to allow third parties to debug nicely with `sys.gettotalrefcount` or `pytest-leaks` (at least if they use object arrays).

seberg committed:
MAINT: Remove need for intermediate refcount incs in dealloc

This enables allocating the simple iterator on the stack using a
borrowed reference to the array. Using this approach during
deallocating object arrays (decrefing the elements) avoids changing
the reference count of `self` during dealloc, which leads to problems.

Just INCREF'ing self is not desirable because it increases the global
reference count in debugging mode so that it makes reference count
debugging harder.

seberg commented:
I pushed an alternative fix for discussion. We could just allocate the iterator object statically (basically using a borrowed reference). For the old style/simple iterator that is fairly simple.

mattip commented:
I think this is fine. Since we should not be calling `PyArray_IterNew` inside a `tp_dealloc`, yet we still have to iterate over all the elements and decref them, it seems "stealing" a reference to a `PyArrayIterObject` this way is really the only way to do it.

seberg committed:
BUG: Fix reference count issue in recursive `dtype` lookup error

seberg committed:
TST: Mark test which increases global reference count

The increase happens in dealloc and is thus harmless. Ignore it,
since it is in a deprecated code path, so downstream should not
be running into this path anyway (the leak is not a real leak,
so that the only reason to avoid is, is to not trip downstream
testing for reference leaks).

seberg commented:
Anyone wishes to chime in? I feel it is fine to create this internal helper (`PyArray_RawIterBaseInit`) to allow creating the old-style iterator as a local stack allocation, and – because it is guaranteed to be local – not need to hold on a reference to the array.

I am also somewhat fine with not doing it, if anyone feels it is a pattern we never use elsewhere and is just a bit too special to start using.

pv reviewed (commented):
+1, fix itself looks necessary. I guess question about names is not important.

seberg commented on numpy/core/src/multiarray/iterators.c:
```diff
@@ -116,10 +116,12 @@ get_ptr_simple(PyArrayIterObject* iter, npy_intp *coordinates)
  */
-static PyObject *
-array_iter_base_init(PyArrayIterObject *it, PyArrayObject *ao)
+void
```
Yes, I think you are right. Changed that (and squashed the commits). I left the name for now, but I do not have much of an opinion. We sometimes use `PyArray_` for internal names, and sometimes not. I do not think we are very consistent about it.

pv replied:
Well, I think it's also ok as it is now.

seberg commented:
@pv thanks, so we can just merge this?

pv commented:
I think so. LGTM, and no smoke came out when using this.

seberg commented:
OK, I will just merge it then. Thanks Pauli!
